### ISSUE TITLE
Error when trying to snapshot disks in Workflow (parallel)

### ISSUE THREAD

**Issue by badc0re** [Azure/azure-powershell#5374]
### Cmdlet(s)
New-AzureRmSnapshot
New-AzureRmSnapshotConfig

### PowerShell Version
PSVersion                      5.1.15063.726

### OS Version
Azure automation script.

### Description
Failed to create managed disk snapshots in parallel.

### Debug Output

> Instructions: to get Debug Output, set `$DebugPreference="Continue"` and then execute the cmdlet or script causing the issue
> Microsoft.PowerShell.Utility\Write-Error : Cannot bind parameter 'Snapshot'. Cannot convert value 
> 
> "Microsoft.Azure.Commands.Compute.Automation.Models.PSSnapshot" to type 
> 
> "Microsoft.Azure.Commands.Compute.Automation.Models.PSSnapshot". Error: "Cannot convert the 
> 
> "Microsoft.Azure.Commands.Compute.Automation.Models.PSSnapshot" value of type 
> 
> "Deserialized.Microsoft.Azure.Commands.Compute.Automation.Models.PSSnapshot" to type 
> 
> "Microsoft.Azure.Commands.Compute.Automation.Models.PSSnapshot"."
> 
> At Snapshot-All:99 char:99
> 
> + 
> 
>     + CategoryInfo          : NotSpecified: (:) [Write-Error], RemoteException
> 
>     + FullyQualifiedErrorId : 
> 
> System.Management.Automation.RemoteException,Microsoft.PowerShell.Commands.WriteErrorCommand
> 
>     + PSComputerName        : [localhost]

### Script/Steps for Reproduction

```
workflow Snapshot-All {
    Param($DisksToSnapshot, $ResourceGroupName, $location, $todaysDate)
    Write-Output ("Start in parallel.")
    
    ForEach -parallel ($dataDiskName in $DisksToSnapshot)
    {    
        Login
        Write-Output ("Snapshotting disk " + $dataDiskName + "..")
        $disk = Get-AzureRmDisk -ResourceGroupName $ResourceGroupName -DiskName $dataDiskName 
        $snapshot =  New-AzureRmSnapshotConfig -SourceUri $disk.Id -CreateOption Copy -Location $location -AccountType PremiumLRS
        $snapshotName = $dataDiskName + "-snapshot" + "-" + $todaysDate
        Write-Output ("Will create snapshot " + $snapshotName)
        New-AzureRmSnapshot -Snapshot $snapshot -SnapshotName $snapshotName -ResourceGroupName $ResourceGroupName
    }
```

**Comment by markcowl** [comment]
@singhkays Can you take a look?  This is one reason why comnplex parameter types can be undesirable in cmdlets

@maddieclayton @markcowl @cormacpayne @sph This is a sample of a type that does not serialize well using the standard powershell serialization system

**Comment by singhkays** [comment]
@markcowl Can you elaborate on the issue? (sorry, I don't know the details of how complex parameters are conflicting here :) ) and what's the recommended approach here?

**Comment by MiYanni** [comment]
@markcowl File a new issue for the general process of determining where else these issues may occur.

**Comment by markcowl** [comment]
@sighkays The issue is that by default, PowerShell limits the depth of Serialization for complex types.  You can update this value in types.ps1xml, as we have done here: https://github.com/Azure/azure-powershell-pr/blob/preview/src/ResourceManager/Profile/Commands.Profile/Microsoft.Azure.Commands.Profile.types.ps1xml#L10

**Comment by singhkays** [comment]
Thanks @markcowl  @hyonholee Can you look into this?

**Comment by markcowl** [comment]
@hyonholee Please let me know if you have any questions about how to do this - this is something your generator could easily handle automatically.

**Comment by Drewm3** [comment]
Cleaning up old issues from 2016 - 2018. Please reopen this issue if it is still a concern.
